### PR TITLE
St 524

### PR DESCRIPTION
**Pull request by vajarova**
- [x] Unit tests passing
- [x] Coverage reached
- [x] Acceptance tests passing
- [ ] Reviewed

### CODE REVIEW

File: app/uk/gov/hmrc/tai/service/journeyCompletion/JourneyCompletionService.scala [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import com.google.inject.Inject
+import com.google.inject.name.Named
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.tai.service.journeyCache.JourneyCacheService
+import uk.gov.hmrc.tai.util.constants.JourneyCacheConstants
+import scala.concurrent.ExecutionContext.Implicits.global
+
+import scala.concurrent.Future
+
+abstract class JourneyCompletionService(successfulJourneyCacheService: JourneyCacheService) extends JourneyCacheConstants {
```
Comment (oc-ste) [inline app/uk/gov/hmrc/tai/service/journeyCompletion/JourneyCompletionService.scala]: Does this need to be an extract class? It's only extended in one place

File: test/controllers/IncomeControllerSpec.scala [hunk 2]
```diff
@@ -337,6 +339,9 @@ class IncomeControllerSpec extends PlaySpec
         when(taxAccountService.updateEstimatedIncome(any(), any(), any(), any())(any())).
           thenReturn(Future.successful(TaiSuccessResponse))
 
+        when(estimatedPayJourneyCompletionService.journeyCompleted(Matchers.eq(employerId.toString))(any())).
+          thenReturn(Future.successful(Map.empty[String, String]))
+
         val result = testController.updateEstimatedIncome()(fakeRequest)
```
Comment (oc-ste) [inline test/controllers/IncomeControllerSpec.scala]: Missing a test if this future fails

File: test/controllers/IncomeControllerSpec.scala [hunk 3]
```diff
@@ -313,6 +312,9 @@ class IncomeControllerSpec extends PlaySpec
         when(taxAccountService.updateEstimatedIncome(any(), any(), any(), any())(any())).
```
Comment (oc-ste) [inline test/controllers/IncomeControllerSpec.scala]: Unrelated to this PR but the toggle on/off expectations are the same?

File: test/controllers/IncomeSourceSummaryControllerSpec.scala [hunk 4]
```diff
@@ -103,7 +113,7 @@ class IncomeSourceSummaryControllerSpec extends PlaySpec
           Future.successful(TaiSuccessResponseWithPayload[Seq[TaxCodeIncome]](taxCodeIncomes)))
         when(employmentService.employment(any(), any())(any())).thenReturn(Future.successful(None))
 
-        val result = sut.onPageLoad(2)(RequestBuilder.buildFakeRequestWithAuth("GET"))
+        val result = sut.onPageLoad(1)(RequestBuilder.buildFakeRequestWithAuth("GET"))
```
Comment (oc-ste) [inline test/controllers/IncomeSourceSummaryControllerSpec.scala]: magic number, is this employmentId or pensionId? It's been changed from 1 -> 2

File: test/controllers/IncomeSourceSummaryControllerSpec.scala [hunk 5]
```diff
@@ -69,12 +76,15 @@ class IncomeSourceSummaryControllerSpec extends PlaySpec
 
       "asked for pension details" in {
         val sut = createSUT
+        val pensionId = 2
         when(taxAccountService.taxCodeIncomes(any(), any())(any())).thenReturn(
           Future.successful(TaiSuccessResponseWithPayload[Seq[TaxCodeIncome]](taxCodeIncomes)))
         when(employmentService.employment(any(), any())(any())).thenReturn(Future.successful(Some(employment)))
         when(benefitsService.benefits(any(), any())(any())).thenReturn(Future.successful(benefits))
+        when(estimatedPayJourneyCompletionService.hasJourneyCompleted(Matchers.eq(pensionId.toString))(any())).
+          thenReturn(Future.successful(true))
```
Comment (oc-ste) [inline test/controllers/IncomeSourceSummaryControllerSpec.scala]: Missing a test for if the future fails

File: test/controllers/income/estimatedPay/update/IncomeUpdateCalculatorControllerSpec.scala [hunk 6]
```diff
@@ -985,6 +986,9 @@ class IncomeUpdateCalculatorControllerSpec
         Future.successful(TaiSuccessResponse)
       )
 
+      when(estimatedPayJourneyCompletionService.journeyCompleted(Matchers.eq(employerId.toString))(any())).
+        thenReturn(Future.successful(Map.empty[String, String]))
+
       val result: Future[Result] = testController.submitIncomeIrregularHours(1)(
```
Comment (oc-ste) [inline test/controllers/income/estimatedPay/update/IncomeUpdateCalculatorControllerSpec.scala]: Missing a test for if the future fails

File: test/uk/gov/hmrc/tai/service/journeyCompletion/EstimatedPayJourneyCompletionServiceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+class EstimatedPayJourneyCompletionServiceSpec extends PlaySpec with MockitoSugar with JourneyCacheConstants with BeforeAndAfterEach {
+
+  private def createTestService = new EstimatedPayJourneyCompletionServiceTest
+
+  val successfulJourneyCacheService: JourneyCacheService = mock[JourneyCacheService]
+  private implicit val hc: HeaderCarrier = HeaderCarrier()
+
+  val incomeId = "1"
+  val trueValue = true.toString
+
```
Comment (oc-ste) [inline test/uk/gov/hmrc/tai/service/journeyCompletion/EstimatedPayJourneyCompletionServiceSpec.scala]: Is this just "true" ?

File: test/uk/gov/hmrc/tai/service/journeyCompletion/EstimatedPayJourneyCompletionServiceSpec.scala [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    "retrieve a cache entry upon request" in {
+      val idKey=s"$TrackSuccessfulJourney_EstimatedPayKey-$incomeId"
+      when(successfulJourneyCacheService.currentValue(meq(idKey))(any())).thenReturn(Future.successful(Some(trueValue)))
+
+      Await.result(createTestService.hasJourneyCompleted(incomeId)(hc), 5 seconds)
+      verify(successfulJourneyCacheService, times(1)).currentValue(meq(idKey))(any())
+    }
+
```
Comment (oc-ste) [inline test/uk/gov/hmrc/tai/service/journeyCompletion/EstimatedPayJourneyCompletionServiceSpec.scala]: Missing the future failed cases

File: app/uk/gov/hmrc/tai/service/journeyCompletion/JourneyCompletionService.scala (hunk 1 above)
Comment (lewey) [reply app/uk/gov/hmrc/tai/service/journeyCompletion/JourneyCompletionService.scala]: Was potentially expecting more classes to come along an extend. The duplicate warning logic could also make use of the new functionality in a refactor

File: test/controllers/IncomeControllerSpec.scala (hunk 3 above)
Comment (lewey) [reply test/controllers/IncomeControllerSpec.scala]: Slightly different in the views that are resolved

**Pull request merged**